### ISSUE TITLE
Unable to provision harvester rke2 cluster from rancher, stuck in provisioning on bootstrap node

### ISSUE THREAD

**Issue by TachunLin** [rancher/rancher#36716]
**Rancher Server Setup**
- Rancher version: v2.6-head (`1d2b746`)
- Installation option (Docker install/Helm Chart): Docker

**Information about the Cluster**
- Kubernetes version: v1.22.6+rke2r1
- Cluster Type (Local/Downstream): Infrastructure Provider - Harvester

**User Information**
- What is the role of the user logged in? Admin
   
**Describe the bug**

Unable to provision RKE2 cluster from rancher `v2.6-head`, keep in `Provisioning` status with message 
This issue exists on both RKE2 `1.22.7+rke2r1` and `1.21.10+rke2r1` 

```
Non-ready bootstrap machine(s) rke2-ubuntu-pool1-bcfd5cfbb-2pj9w and join url to be available on bootstrap node
```

![image](https://user-images.githubusercontent.com/29251855/156367737-43717d56-b5be-43ac-be52-724a2cc3cd57.png)

Further checking cluster node status in RKE2 VM, it looks like node have taints to prevent component to finish deployment.

```
# ssh to rke2 virtual machine

export KUBECONFIG=/etc/rancher/rke2/rke2.yaml
/var/lib/rancher/rke2/bin/kubectl get nodes -o yaml
```

```
spec:
    podCIDR: 10.42.0.0/24
    podCIDRs:
    - 10.42.0.0/24
    taints:
    - effect: NoSchedule
      key: node.cloudprovider.kubernetes.io/uninitialized
      value: "true"
```

**To Reproduce**
 
1. Create a one node or three nodes harvester cluster 
2. Prepare SLES JeOS image
`SLES15-SP3-JeOS.x86_64-15.3-OpenStack-Cloud-GM.qcow2`
4. Enable virtual network to `harvester-mgmt`
5. Create virtual network `vlan1` with id `1`
6. Import harvester in rancher v2.6-head
7. Create cloud credential 
8. Provision a RKE2 cluster with SLES JeOS image  
![image](https://user-images.githubusercontent.com/29251855/156369683-d94ba2be-1ab5-4df8-9328-013dda470583.png)
9. Check the RKE2 cluster provisioning to `Ready`

**Result**
Provisioning stuck in `Non-ready bootstrap machine(s) rke2-ubuntu-pool1-bcfd5cfbb-2pj9w and join url to be available on bootstrap node`

**Expected Result**

Can provision RKE2 cluster in harvester correctly in acceptable time for major OS image (e.g SLES, Ubuntu)

**Additional context**

- Harvester support bundle
  [supportbundle_74b92604-c442-4d1d-a70d-7313e32ac47c_2022-03-02T13-14-35Z.zip](https://github.com/harvester/harvester/files/8170323/supportbundle_74b92604-c442-4d1d-a70d-7313e32ac47c_2022-03-02T13-14-35Z.zip)

 - Rancher container log
  [rancher2.6-head_rke2_sles.log](https://github.com/harvester/harvester/files/8170358/rancher2.6-head_rke2_sles.log)

**Comment by TachunLin** [comment]
After double checking, we can provision RKE2 cluster on latest rancher `v2.6-head` (33db2f5) with `1.21.10+rke2r1`
Currently still failed to provision RKE2 on `1.22.7+rke2r1`. 

### Test Envrionment

- Harvester ISO version: master-088a020f-head (03/05)
- Underlying Infrastructure: 1 node harvester on local kvm machine
- Rancher version: v2.6-head (33db2f5)

**Comment by thedadams** [comment]
@TachunLin Is this only a problem with SLES? Are you able to provision using Ubuntu image?

**Comment by TachunLin** [comment]
@thedadams, The issue exists on both SLES and Ubuntu image. 
The previous check result https://github.com/rancher/rancher/issues/36716#issuecomment-1059969652 was based on Ubuntu cloud image.

**Comment by TachunLin** [comment]
@thedadams ,we double confirmed currently the RKE2 provisioning status according to https://github.com/harvester/harvester/issues/1823#issuecomment-1049501074

|       | 1.21.10+rke2r1    | 1.22.7+rke2r1                         |
|-------|-------------------|---------------------------------------|
| Focal | Work              | RKE2 starts but has NoSchedule taint. |
| JeOS  | etcd can't start. | RKE2 starts but has NoSchedule taint. |

1. On `1.22.7+rke2r1`, **can't** provision RKE2 cluster with both `Ubuntu cloud focal` and `SLES JeOS` images
2. On `1.21.10+rke2r1`, **can** provision RKE2 cluster on `Ubuntu focal`
3. On `1.21.10+rke2r1`, **can't** provision RKE2 cluster on `SLES JeOS`, but with manual install apparmor-parser can workaround it

For SLES JeOS provisioning, we create and track in the separate issue https://github.com/harvester/harvester/issues/1973
And track the Ubuntu focal issue on https://github.com/harvester/harvester/issues/1954

**Comment by thedadams** [comment]
@TachunLin The SLES issue seems to be unrelated to the issue with Ubuntu. Rancher is not responsible for installing the apparmor-parser, so the SLES issue is understood (the installation of apparmor-parser could be added to rancher-machine when the SLES operating system is detected).

For the Ubuntu image case, the nodes are tainted as `NoSchedule` because the cloud-provider (Harvester, in this case) isn't setup on the node. Has it been verified that the Harvester cloud-provider is running successfully on these nodes? And has the Harvester cloud-provider been tested on Kubernetes v1.22+ outside of Rancher provisioning?

**Comment by TachunLin** [comment]
@thedadams, we haven't tested the harvester cloud-provider outside of Rancher provisioning since the verification process in Harvester are integrated with Rancher. 
And we previously tested the RKE2 provision on `1.22.6`. There we can provision RKE2 on rancher `v2.6.4-rc3`  with harvester master-head release after v1.0-GA with Ubuntu cloud image (as the test result in https://github.com/harvester/harvester/issues/1659#issuecomment-1041517272)

**Comment by thedadams** [comment]
@TachunLin Thanks for that information. That is helpful.

I am still wondering if the cloud provider is running on the Ubuntu nodes with `1.22.7+rke2r1`. It seems to me that the cloud provider not running on the node would be the most likely reason that the taint isn't getting removed.

**Comment by FrankYang0529** [comment]
Hi @thedadams, cloud provider is running, but there is some error message in it.
```
node_controller.go:390] Initializing node rke2-in-jeos-pool1-51a1e43e-fm4kk with cloud provider
node_controller.go:212] error syncing 'rke2-in-jeos-pool1-51a1e43e-fm4kk': failed to get instance metadata for node rke2-in-jeos-pool1-51a1e43e-fm4kk: Get "https://192.168.0.131:6443/apis/kubevirt.io/v1/namespaces/default/virtualmachines/rke2-in-jeos-pool1-51a1e43e-fm4kk": dial tcp 192.168.0.131:6443: i/o timeout, requeuing
```

**Comment by thedadams** [comment]
Thanks @FrankYang0529 for looking into that. Is that a Rancher issue then? Or would this be an issue with the Harvester cloud provider?

**Comment by FrankYang0529** [comment]
Let me investigate harvester-cloud-provider. Thank you.

**Comment by FrankYang0529** [comment]
@thedadams, it looks like the root cause is `Calico` can't be installed. If we change CNI to `Canal`, RKE2 `1.22.7+rke2r1` can be installed.

In RKE2 `1.22.7+rke2r1`, it upgrades `tigera-operator` to `v1.23.5`. Also, `tigera-operator` updates tolerations in `calico-typha`.  Originally, in `tigera-operator v1.17.6`, it use tolerations for `calico-typha` like following, so we can install RKE2 `1.21.10+rke2r1` without error.
      tolerations:
      - key: CriticalAddonsOnly
        operator: Exists
      - effect: NoSchedule
        operator: Exists
      - effect: NoExecute
        operator: Exists
However, in tigera-operator v1.23.5, it adds keys to tolerations like following, so calico-typha can't tolerate [node.cloudprovider.kubernetes.io/uninitialized](http://node.cloudprovider.kubernetes.io/uninitialized).
      tolerations:
      - effect: NoSchedule
        key: node-role.kubernetes.io/control-plane
        operator: Exists
      - effect: NoExecute
        key: node-role.kubernetes.io/etcd
        operator: Exists

**Comment by snasovich** [comment]
@manuelbuil , is this something that needs to be fixed on RKE2 side itself? If so, could we transfer the issue there?

**Comment by manuelbuil** [comment]
I thought this was fixed already. The harvester team edited the cluster YAML to add the required toleration: https://suse.slack.com/archives/C02D01ARDL5/p1646736123552459?thread_ts=1646638768.682839&cid=C02D01ARDL5

**Comment by FrankYang0529** [comment]
Hi @manuelbuil, if `calico-typha` can't tolerate `node.cloudprovider.kubernetes.io/uninitialized`, this may also be an issue for other cloud providers. If we can fix it as default chart values, we don't need to modify Cluster YAML each time.

**Comment by manuelbuil** [comment]
Is calico the only component that requires this toleration? What about others like coredns?

**Comment by FrankYang0529** [comment]
> Is calico the only component that requires this toleration? What about others like coredns?

Sorry for the late reply, I was doing some testing. `coredns` is also pending cause of the taint, but the cluster can start after I change `rke2-calico` to the following:

```
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-calico
  namespace: kube-system
spec:
  valuesContent: |-
    installation:
      controlPlaneTolerations:
      - key: "node-role.kubernetes.io/control-plane"
        operator: "Exists"
        effect: "NoSchedule"
      - key: "node-role.kubernetes.io/etcd"
        operator: "Exists"
        effect: "NoExecute"
      - key: "node.cloudprovider.kubernetes.io/uninitialized"
        operator: "Exists"
        effect: "NoSchedule"
```

**Comment by manuelbuil** [comment]
> > Is calico the only component that requires this toleration? What about others like coredns?
> 
> Sorry for the late reply, I was doing some testing. `coredns` is also pending cause of the taint, but the cluster can start after I change `rke2-calico` to the following:
> 
> ```
> apiVersion: helm.cattle.io/v1
> kind: HelmChartConfig
> metadata:
>   name: rke2-calico
>   namespace: kube-system
> spec:
>   valuesContent: |-
>     installation:
>       controlPlaneTolerations:
>       - key: "node-role.kubernetes.io/control-plane"
>         operator: "Exists"
>         effect: "NoSchedule"
>       - key: "node-role.kubernetes.io/etcd"
>         operator: "Exists"
>         effect: "NoExecute"
>       - key: "node.cloudprovider.kubernetes.io/uninitialized"
>         operator: "Exists"
>         effect: "NoSchedule"
> ```

Good. But your request is that calico includes this toleration by default, right?

**Comment by FrankYang0529** [comment]
> Good. But your request is that calico includes this toleration by default, right?

Yes

**Comment by manuelbuil** [comment]
According to what I learnt (thanks @FrankYang0529 ), the harvester-cloud-provider is the one removing the taint `node.cloudprovider.kubernetes.io/uninitialized` once it considers the node is ready. That provider fails with the error:
```
node_controller.go:390] Initializing node rke2-in-jeos-pool1-51a1e43e-fm4kk with cloud provider
node_controller.go:212] error syncing 'rke2-in-jeos-pool1-51a1e43e-fm4kk': failed to get instance metadata for node rke2-in-jeos-pool1-51a1e43e-fm4kk: Get "https://192.168.0.131:6443/apis/kubevirt.io/v1/namespaces/default/virtualmachines/rke2-in-jeos-pool1-51a1e43e-fm4kk": dial tcp 192.168.0.131:6443: i/o timeout, requeuing
```
And thus it is unable to remove the taint.

I am confused about why it requires the CNI plugin to be running. The url `https://192.168.0.131:6443/apis/....` looks like a call to kube-api and kube-api is running in `hostNetwork: true`, i.e. it does not require a cni plugin to run. Could you verify if `192.168.0.131` is the IP where kube-api is running?

**Comment by yaocw2020** [comment]
> node.cloudprovider.kubernetes.io/uninitialized

@manuelbuil Yes, good consideration! We should enable the hostNetwork in the Harvester cloud provider to solute this issue completely.

**Comment by manuelbuil** [comment]
> > node.cloudprovider.kubernetes.io/uninitialized
> 
> @manuelbuil Yes, good consideration! We should enable the hostNetwork in the Harvester cloud provider to solute this issue completely.

I was assuming harvester-cloud-provider was using hostNetwork mode. If it is not, then it would not run because it would require the CNI plugin to be ready. But it is running, right?

**Comment by yaocw2020** [comment]
> 

Yes. I guess the CNI tolerates the taint `node.cloudprovider.kubernetes.io/uninitialized` if it is running.

**Comment by manuelbuil** [comment]
> > > > node.cloudprovider.kubernetes.io/uninitialized
> > > 
> > > 
> > > @manuelbuil Yes, good consideration! We should enable the hostNetwork in the Harvester cloud provider to solute this issue completely.
> > 
> > 
> > I was assuming harvester-cloud-provider was using hostNetwork mode. If it is not, then it can't not run because it requires the CNI plugin to be ready. But it is running, right?
> 
> Yes. I guess the CNI tolerates the taint `node.cloudprovider.kubernetes.io/uninitialized` if it is running.

Could it be that this happens?:
1 - k8s Cluster starts
2 - There is no taint so calico starts
3 - Once calico is started, `harvester-cloud-provider` pod is scheduled and starts
4 - harvester-cloud-provider taints `node.cloudprovider.kubernetes.io/uninitialized`
5 - Calico stops working because of the taint
6 - harvester-cloud-provider can't access harvester kube-api outside of the cluster

**Comment by yaocw2020** [comment]
@manuelbuil No, the taint is added by the kubelet once we set the --cloud-provider as external according the kubernetes document https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/

**Comment by FrankYang0529** [comment]
> 2 - If typha is showing problems, why calico-node is not? Both shouldn't be tolerating the taint

@manuelbuil here is tolerations in `calico-node`.

```
  tolerations:
  - key: CriticalAddonsOnly
    operator: Exists
  - effect: NoSchedule
    operator: Exists
  - effect: NoExecute
    operator: Exists
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
  - effect: NoSchedule
    key: node.kubernetes.io/disk-pressure
    operator: Exists
  - effect: NoSchedule
    key: node.kubernetes.io/memory-pressure
    operator: Exists
  - effect: NoSchedule
    key: node.kubernetes.io/pid-pressure
    operator: Exists
  - effect: NoSchedule
    key: node.kubernetes.io/unschedulable
    operator: Exists
  - effect: NoSchedule
    key: node.kubernetes.io/network-unavailable
    operator: Exists
```

**Comment by manuelbuil** [comment]
> > 2 - If typha is showing problems, why calico-node is not? Both shouldn't be tolerating the taint
> 
> @manuelbuil here is tolerations in `calico-node`.
> 
> ```
>   tolerations:
>   - key: CriticalAddonsOnly
>     operator: Exists
>   - effect: NoSchedule
>     operator: Exists
>   - effect: NoExecute
>     operator: Exists
>   - effect: NoExecute
>     key: node.kubernetes.io/not-ready
>     operator: Exists
>   - effect: NoExecute
>     key: node.kubernetes.io/unreachable
>     operator: Exists
>   - effect: NoSchedule
>     key: node.kubernetes.io/disk-pressure
>     operator: Exists
>   - effect: NoSchedule
>     key: node.kubernetes.io/memory-pressure
>     operator: Exists
>   - effect: NoSchedule
>     key: node.kubernetes.io/pid-pressure
>     operator: Exists
>   - effect: NoSchedule
>     key: node.kubernetes.io/unschedulable
>     operator: Exists
>   - effect: NoSchedule
>     key: node.kubernetes.io/network-unavailable
>     operator: Exists
> ```

`node.cloudprovider.kubernetes.io/uninitialized` is not part of it. Can you confirm that calico-node is not running?

**Comment by FrankYang0529** [comment]
> `node.cloudprovider.kubernetes.io/uninitialized` is not part of it. Can you confirm that calico-node is not running?

Yeah, but it has `NoSchedule` without key, so it can tolerate `node.cloudprovider.kubernetes.io/uninitialized`. `calico-node` is running, but it's not ready. It has error message like following.
```
calico-node 2022-03-07 07:35:24.461 [ERROR][79245] tunnel-ip-allocator/discovery.go 153: Didn't find any ready Typha instances.
calico-node 2022-03-07 07:35:24.461 [FATAL][79245] tunnel-ip-allocator/startsyncerclient.go 48: Typha discovery enabled but discovery failed. error=Kubernetes service missing IP or port
```

**Comment by snasovich** [comment]
It doesn't seem to be related to RKE2 cluster provisioning itself, so I'm removing `team/area2` label as it's not something our team will be fixing. Please let me know if that's not the case (@-mention me, please).

**Comment by TachunLin** [comment]
Verified **fixed** on  on Rancher `v2.6-head 2642acc` (03/16) with Harvester `v1.0-591d8015-head` (03/16). Close this issue.

### Result
Can provision RKE2 cluster with kubernetes version **`v1.21.7+rke2r2`** for the following OS image:
1. Ubuntu focal image
2. SLES JeOS image

![image](https://user-images.githubusercontent.com/29251855/158577116-0bda5f89-2083-49e9-8f55-d6d172df05ef.png)
![image](https://user-images.githubusercontent.com/29251855/158577306-996c53dd-a7e1-43b2-af59-97896faf47f8.png)

### Test Information
* Test Environment: 1 nodes harvester on local kvm machine
* Harvester version: v1.0-591d8015-head (03/16)
* Rancher version: v2.6-head 2642acc
* Kubernetes version: v1.21.7+rke2r2
* OS image: Ubuntu focal, SLES JeOS

### Verify Steps
1. Create a one node harvester 
2. Upload Ubuntu focal and SLES JeOS image
3. Create virtual network vlan1 with id 1
4. Import Harvester from Rancher
5. Create cloud credential 
6. Provision an RKE2 cluster 
7. Add the following use cloud-init data 
```
  password: 123456
  chpasswd: { expire: False }
  ssh_pwauth: True
  ```
8. Select Ubuntu focal or SLES JeOS image
9. Provide available cpu and memory resources
10. Provision the RKE2 cluster

**Comment by TachunLin** [comment]
@manuelbuil, this issue have been verified fixed as above test result. Please help to change the status accordingly. Thank you.

**Comment by lanfon72** [comment]
update another case which also caused same error:

Create Cluster using **External Harvester** (rke2)
The `kubeconfig` will not be requested, and the cluster lost some yaml field.
![image](https://user-images.githubusercontent.com/5169694/158817364-15f455d2-d64c-4f25-9d27-a375cb2d8989.png)
![image](https://user-images.githubusercontent.com/5169694/158817759-a76cd51a-ee07-4f56-820a-b075370aac6d.png)
![image](https://user-images.githubusercontent.com/5169694/158819226-79b981ca-e14c-462d-9ff2-65e7b7821240.png)


## The Success case
Create Cluster using **Imported Harvester** (rke2)
The `kubeconfig` will be requested, and the cluster contains `cloud-provider-config` field in yaml.
![image](https://user-images.githubusercontent.com/5169694/158818388-2777ddf6-eb64-4051-bda9-67ac6584b9d9.png)
![image](https://user-images.githubusercontent.com/5169694/158819099-c78944c0-d3a5-4b5b-951d-96c948c0c8ba.png)

**Comment by snasovich** [comment]
Closing it as it was supposed to be closed a few days ago and got reopened due to Zube requiring "Done" label on issues.
@TachunLin, I see a comment above by @lanfon72 that this may still be an issue or there is another related issue, please handle it as necessary.
